docs/reference-guide/modules/queries/pages/query-handlers.adoc

```diff
@@ -55,52 +78,101 @@ From the above sample, we want to highlight two specifics when it comes to writi
+. If the `@Query` annotation is present on the query payload class, it defines the qualified name and version of the query.
+. Otherwise, the fully qualified class name of the query payload class is used as the query name.
+
+=== Matching handlers to queries
+
+Query handlers are matched to queries based on the query name. The `@QueryHandler` annotation can optionally specify a `queryName` attribute to explicitly declare which query name the handler wants to handle. If not specified, the handler matches queries whose name corresponds to the fully qualified class name of the handler's first parameter type.
+
+Example using the `@Query` annotation:
 
 [source,java]
 ----
-// assume QueryB extends QueryA
-// and    QueryC extends QueryB
-// and    a single instance of SubHandler is registered
+import org.axonframework.messaging.queryhandling.annotation.Query;
 
-public class QueryHandler {
+@Query(namespace = "giftcard", name = "FetchCardSummary", version = "1.0")
+public class FetchCardSummaryQuery {
+    private final String cardSummaryId;
 
-    @QueryHandler
-    public MyResult handle(QueryA query) {
-        // Return result
+    public FetchCardSummaryQuery(String cardSummaryId) {
+        this.cardSummaryId = cardSummaryId;
     }
 
-    @QueryHandler
-    public MyResult handle(QueryB query) {
-        // Return result
+    public String getCardSummaryId() {
+        return cardSummaryId;
     }
+}
+----
 
-    @QueryHandler
-    public MyResult handle(QueryC query) {
-        // Return result
+The query handler can then be written to handle this specific query type:
 
-    }
+[source,java]
+----
+@QueryHandler
+public CardSummary handle(FetchCardSummaryQuery query) {
+    return cardSummaryStorage.get(query.getCardSummaryId());
 }
+----
 
-public class SubQueryHandler extends QueryHandler {
+Since the handler's first parameter is `FetchCardSummaryQuery`, and that class is annotated with `@Query(namespace = "giftcard", name = "FetchCardSummary")`, the handler will match queries with the qualified name `giftcard.FetchCardSummary`.
 
-    @QueryHandler
-    public MyResult handleEx(QueryB query) {
-        // Return result
-    }
+=== Using queryName for different representations
+
+If you want the handler to receive the query in a different representation (not the `@Query` annotated class), you **must** explicitly specify the query name in the `@QueryHandler` annotation:
+
+[source,java]
+----
+@QueryHandler(queryName = "giftcard.FetchCardSummary") // <1>
+public CardSummary handle(Map<String, Object> queryData) { // <2>
+    String cardId = (String) queryData.get("cardSummaryId");
+    return cardSummaryStorage.get(cardId);
+}
+----
+<1> The `queryName` attribute is **required** when the first parameter is not the `@Query` annotated payload class. Without it, Axon wouldn't know which query this handler should handle.
```

Comment:
This makes it seem as if the `@Query` annotation is a hard requirement, which isn't true. This is completely dependent on the chosen `MessageTypeResolver`. I think we should adjust the wording here to be more inline with the options people have in this area.